### DISCUSSION TITLE
Ubuntu specific, beginner friendly build guide

### DISCUSSION THREAD

**Discussion by somekid6** [NoahGWood/OpenEFT#26] [General]
This guide assumes you have [a fresh copy of Ubuntu 22.04.2 LTS Desktop installed](https://ubuntu.com/download/desktop) and is intended for a non-Ubuntu user audience.

I've compiled everyone's work so far and tried to make it as beginner friendly as I can. I'm not a developer so this might not have the best practices but it does seem to work. I tested it on a fresh install of Ubuntu 22.04.2 LTS Desktop on a dedicated computer (VM + checkpoints for dev.) and made sure there weren't any dependencies missed.

Starting from the desktop, click "Activities" in the upper left of the screen and open "Terminal".

type 

	sudo apt-get update -y && sudo apt-get upgrade -y

 and press enter.

type 

	sudo apt install -y git cmake python3-pip libx11-dev

 and press enter.

type 

	cd ~/

and press enter.

type 

	git clone --recurse-submodules https://github.com/NoahGWood/OpenEFT.git

and press enter.

Click "Activities" in the top left, double click "Files", click "Home", double click "OpenEFT", double click "requirements.txt". 

Change:

	backports.zoneinfo==0.2.1
	
To:

	backports.zoneinfo;python_version<"3.9"
	
Save and close the window.

Now double click to open the file "build_linux.sh".

Copy the script below and replace everything in the file with it, then save/close the file.
	

	#!/usr/bin/bash
	
	mkdir build
	mkdir -p browser/linux
	
	echo "Setting Up NBIS"
	sudo rm -rf ./nbis #Remove any old copies
	git clone https://github.com/OwenPemberton/nbis ./nbis
	cd nbis
	./setup.sh ${PWD}/../build --64
	echo "Configuring NBIS"
	make config
	echo "Making NBIS"
	make it
	echo "Installing NBIS"
	sudo make install LIBNBIS=no
	
	echo "Checking NBIS binaries paths"
	case :$PATH:
	  in *:$HOME/OpenEFT/nbis/nfseg/bin:*) ;; # do nothing, it's there
	     *) echo 'export PATH="$HOME/OpenEFT/nbis/nfseg/bin:$PATH"' >> ~/.bashrc #it's missing, add to PATH
	esac
 
	case :$PATH:
	  in *:$HOME/OpenEFT/nbis/nfiq/bin:*) ;; # do nothing, it's there
	     *) echo 'export PATH="$HOME/OpenEFT/nbis/nfiq/bin:$PATH"' >> ~/.bashrc #it's missing, add to PATH
	esac

	#source ~/.bashrc doesn't work in .sh in Ubuntu since scripts run in their own session. Need to run post build OR run the script with 'source'
	source ~/.bashrc

	echo "Installing UnGoogled Chromium"
	cd ../browser/linux
	wget "https://github.com/clickot/ungoogled-chromium-binaries/releases/download/111.0.5563.65-1/ungoogled-chromium_111.0.5563.65-1.1.AppImage"
	mv ungoogled-chromium_111.0.5563.65-1.1.AppImage chrome.AppImage
	chmod +x chrome.AppImage
	cd ../../
	
	echo "Installing LibOpenJP2-Tools"
	sudo apt-get install libopenjp2-tools -y
	
	echo "Installing Pip Requirements"
	cd ~/OpenEFT
	pip3 install -r requirements.txt
	
	python3 manage.py migrate
	
	echo "OpenEFT build complete. Run the app by typing 'python3 openeft.py'"
	cd ~/OpenEFT

Go back to your terminal window and type 

	cd ~/OpenEFT

 and press enter.

Now type 

	source build_linux.sh

 and press enter. Input your password when it asks. 

Now type 

	python3 openeft.py

 and press enter to start the program

edit - Did some testing and it looks like OpenJPEG isn't needed since moving libopenjp2-tools out of the if statement from the original so I removed that.

edit 2 - updated with fix for Chromium not opening

edit 3 - figure out how to make the script update $PATH of the parent shell process.

edit - the eForm system will be unhappy if you submit the eft on the same day you made it. just wait a day and it'll be happy again.

**Comment by Hyrb69** [comment]
THANK YOU for this effort!
